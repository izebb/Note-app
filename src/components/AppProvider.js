import React, { createContext, Component } from "react";
import { withRouter } from "react-router-dom";

import { Notes } from "../utils/notes";

const AppContext = React.createContext();

export class Provider extends Component {
  constructor(props) {
    super(props);

    this.notes = new Notes();

    this.state = {
      notes: this.getNotes(),
      textEditorInput: "",
      noteTitle: "",
      searchValue: "",
      getNoteById: this.getNoteById,
      onSearch: this.onSearch,
      createNote: this.createNote,
      onTitleChange: this.onTitleChange,
      onEditorChange: this.onEditorChange,
      deleteNote: this.deleteNote,
      updateNote: this.updateNote,
      addNote: this.addNote
    };
  }

  getNotes() {
    return [...this.notes.get()].reverse();
  }

  deleteNote = id => {
    this.notes.delete(id);
    const notes = this.getNotes();

    this.setState({ notes }, this.redirectToNote);
  };

  updateNote = id => {
    const { textEditorInput, noteTitle } = this.state;

    if (noteTitle.trim()) {
      this.notes.update(id, {
        noteTitle,
        textEditorInput
      });

      this.setState({
        notes: this.getNotes()
      });
    }
  };

  createNote = () => {
    const { textEditorInput, noteTitle } = this.state;
    const id = this.notes.create({
      noteTitle: "New note",
      textEditorInput: ""
    });

    this.setState(
      {
        searchValue: "",
        notes: this.getNotes()
      },
      () => this.redirectToNote(id)
    );
  };

  redirectToNote = (id = "") => {
    const { history } = this.props;

    history.push(`/${id}`);
  };

  onEditorChange = e => {
    this.setState({
      textEditorInput: e.target.value
    });
  };

  onSearch = e => {
    const value = e.target.value;

    this.setState({
      searchValue: e.target.value,
      notes: this.notes.search(value.trim())
    });
  };

  onTitleChange = e => {
    this.setState({
      noteTitle: e.target.value
    });
  };

  getNoteById = id => {
    const item = this.notes.get(id);

    if (item) {
      this.setState(item);
    } else {
      this.redirectToNote();
    }
  };

  render() {
    return (
      <AppContext.Provider value={{ ...this.state }} {...this.props}>
        {this.props.children}
      </AppContext.Provider>
    );
  }
}

const AppConsumer = AppContext.Consumer;
const AppProvider = withRouter(Provider);

export { AppContext, AppConsumer, AppProvider };
