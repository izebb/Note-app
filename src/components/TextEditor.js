import React from "react";
import styled from "styled-components";

const EditorContainerStyled = styled.div`
  flex: 1;
`;

const TextArea = styled.textarea`
  width: 100%;
  box-sizing: border-box;
  border: none;
  height: 100%;
  background: transparent;
  font-size: 16px;
  line-height: 1.6;
  font-weight: 300;
  padding: 20px;
  white-space: pre-wrap;
  resize: none;

  &:focus {
    outline: none;
  }
`;

export function TextEditor({ onEditorChange, save, textEditorInput }) {
  return (
    <EditorContainerStyled>
      <TextArea
        onBlur={save}
        value={textEditorInput}
        onChange={onEditorChange}
      />
    </EditorContainerStyled>
  );
}
