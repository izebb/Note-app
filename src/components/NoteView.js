import React, { Fragment, useContext, useEffect } from "react";
import styled from "styled-components";

import { AppContext } from "./AppProvider";
import { TextEditor } from "./TextEditor";

const ViewPaperContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  border: 1px solid #e6e6e6;
  position: relative;
  background-color: rgba(244, 255, 254, 0.78);

  &:before {
    content: " ";
    position: absolute;
    width: 100%;
    height: 100%;
    border: 1px solid #efefef;
    top: 0;
    z-index: -1;
    box-shadow: 1px 2px 10px #f6fffe;
  }
`;

const ViewTitleBarStyled = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const NoteTitleStyled = styled.input`
  width: 100%;
  height: 58px;
  padding: 7px 10px;
  font-size: 20px;
  font-weight: 700;
  background: transparent;
  border: none;
  border-bottom: 1px solid #90b7b4;
  &:focus {
    outline: none;
    border-bottom: 1px solid #5f7d8c;
  }
`;

export function NoteView({ match }) {
  const {
    saveNote,
    noteTitle,
    onTitleChange,
    textEditorInput,
    onEditorChange,
    getNoteById,
    deleteNote,
    updateNote
  } = useContext(AppContext);
  const { id = "" } = match.params;

  useEffect(() => {
    if (id) {
      getNoteById(id);
    }
  }, [id]);

  return (
    <ViewPaperContainer>
      <ViewTitleBarStyled>
        <NoteTitleStyled
          onBlur={() => updateNote(id)}
          value={noteTitle}
          placeholder="Title of note"
          onChange={onTitleChange}
        />
      </ViewTitleBarStyled>
      <TextEditor
        save={() => updateNote(id)}
        onEditorChange={onEditorChange}
        textEditorInput={textEditorInput}
      />
    </ViewPaperContainer>
  );
}
