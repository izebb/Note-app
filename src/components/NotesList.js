import React, { useContext } from "react";
import { NavLink } from "react-router-dom";
import styled from "styled-components";
import dateFormat from "dateFormat";

import binIcon from "../images/bin.svg";
import { AppContext } from "./AppProvider";

const NoteListContainerStyled = styled.div`
  flex: 1;
  overflow-y: auto;
`;

const DeleteButton = styled.button`
  background-color: transparent;
  border: none;
  width: 40px;
  position: absolute;
  right: 5px;
  opacity: 0;
  transform: translateX(20px);
  transition: transform 200ms ease-out, opacity 150ms ease-out;
`;

const NoteItemStyled = styled.div`
  display: flex;
  flex-direction: column;
  padding: 15px 20px 12px 20px;
  margin-left: 40px;
  color: #ccc;
  border-bottom: 1px dashed #b5b5b5;
  text-decoration: none;
  position: relative;
  padding-right: 45px;
  overflow: hidden;

  &:hover {
    ${DeleteButton} {
      opacity: 0.8;
      transform: none;
    }
  }
  &.selected {
    color: #fff;
  }
`;

const NoteItemTitleStyled = styled.div`
  font-weight: 700;
`;

const NoteItemDateStyled = styled.div`
  font-size: 12px;
  margin-top: 20px;
`;

const SearchContainerStyled = styled.div`
  padding: 0 30px;
  margin: 10px 0;
`;

const SearchStyled = styled.input`
  width: 100%;
  height: 30px;
  box-sizing: border-box;
  border: none;
  padding: 0 5px;
  border-radius: 4px;
  font-size: 14px;

  &:focus {
    outline: none;
  }
`;

export function NotesList({ searchValue, onSearch, data, onClick, onDelete }) {
  const removeNote = id => () => {
    if (confirm("Do you really wanna delete this note")) {
      onDelete(id);
    }
  };

  return (
    <NoteListContainerStyled>
      <SearchContainerStyled>
        <SearchStyled
          value={searchValue}
          onChange={onSearch}
          placeholder="Search Notes..."
        />
      </SearchContainerStyled>
      {data.map(item => {
        return (
          <NoteItemStyled
            className="note"
            as={NavLink}
            key={item.id}
            to={`/${item.id}`}
            activeClassName="selected"
          >
            <NoteItemTitleStyled>{item.noteTitle}</NoteItemTitleStyled>
            <NoteItemDateStyled>
              {dateFormat(item.createdAt, "mmmm dS, yyyy, h:MM:ss TT")}
            </NoteItemDateStyled>
            <DeleteButton id="delete_button" onClick={removeNote(item.id)}>
              <img src={binIcon} />
            </DeleteButton>
          </NoteItemStyled>
        );
      })}
    </NoteListContainerStyled>
  );
}

NotesList.defaultProps = {
  data: [],
  onClick: () => {},
  onDelete: () => {},
  onSearch: () => {}
};
