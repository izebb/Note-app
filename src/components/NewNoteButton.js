import React from "react";
import styled from "styled-components";
import { Link } from "react-router-dom";

import { NotesList } from "./NotesList";

const NewNoteStyled = styled.div`
  width: 100%;
  height: 100px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const NewButtonStyled = styled.button`
  color: #fff;
  background-color: #6021f3;
  padding: 0 50px;
  height: 37px;
  font-size: 13px;
  border-radius: 5px;
  height: 37px;
  font-size: 13px;
  border-radius: 50px;
  box-shadow: 1px 1px 10px rgba(0, 0, 0, 0.32);
  border: none;
  font-weight: 600;
  text-decoration: none;
  display: flex;
  align-items: center;
  &:focus {
    text-decoration: none;
  }
`;

export function NewNoteButton({ onClick }) {
  return (
    <NewNoteStyled>
      <NewButtonStyled onClick={onClick}>Add Note</NewButtonStyled>
    </NewNoteStyled>
  );
}

NewNoteButton.defaultProps = {
  onClick: () => {}
};
