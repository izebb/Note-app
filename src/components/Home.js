import React, { useContext } from "react";
import { Redirect } from "react-router-dom";
import styled from "styled-components";
import { AppContext } from "./AppProvider";
import { NewNoteButton } from "./NewNoteButton";

import journalImage from "../images/journal.svg";

const ImgStyled = styled.img`
  height: 240px;
  opacity: 0.35;
`;

const TextStyled = styled.p`
  color: #444;
  margin-bottom: 20px;
`;

const EmptyContainerStyled = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  flex: 1;
  font-weight: 300;
  font-size: 30px;
  padding-top: 130px;
  color: #444;
`;

export function Home() {
  const { notes, createNote } = useContext(AppContext);
  const item = notes[0];

  if (item) {
    return <Redirect to={`/${item.id}`} replace />;
  }

  return (
    <EmptyContainerStyled>
      <ImgStyled src={journalImage} />
      <TextStyled>You do not have any notes. Let's make one for you</TextStyled>
      <div>
        <NewNoteButton onClick={createNote} />
      </div>
    </EmptyContainerStyled>
  );
}
