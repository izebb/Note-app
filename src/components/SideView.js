import React, { useContext } from "react";
import styled from "styled-components";

import { AppContext } from "./AppProvider";
import { NotesList } from "./NotesList";
import { NewNoteButton } from "./NewNoteButton";

const SideViewStyled = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  width: 300px;
  height: 100%;
  background-color: #576cff;
  padding: 15px 0 0 0;
  display: flex;
  flex-direction: column;
`;

export function SideView() {
  const { notes, createNote, deleteNote, searchValue, onSearch } = useContext(
    AppContext
  );

  return (
    <SideViewStyled>
      <NewNoteButton onClick={createNote} />
      <NotesList
        data={notes}
        onDelete={deleteNote}
        onSearch={onSearch}
        searchValue={searchValue}
      />
    </SideViewStyled>
  );
}
