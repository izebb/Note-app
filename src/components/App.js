import React from "react";
import styled from "styled-components";
import { BrowserRouter as Router, Route } from "react-router-dom";

import { SideView } from "./SideView";
import { NoteView } from "./NoteView";
import { AppProvider } from "./AppProvider";
import { Home } from "./Home";

const AppContainerStyled = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  min-width: 1200px;
  width: 100%;
  height: 100%;
`;

const MainViewStyled = styled.div`
  height: 100%;
  margin-left: 300px;
  display: flex;
`;

const MainViewContainer = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  max-width: 750px;
  margin: 30px auto;
`;

export function App() {
  return (
    <AppContainerStyled>
      <Router>
        <AppProvider>
          <SideView />
          <MainViewStyled>
            <MainViewContainer>
              <Route path="/" exact component={Home} />
              <Route path="/:id" component={NoteView} />
            </MainViewContainer>
          </MainViewStyled>
        </AppProvider>
      </Router>
    </AppContainerStyled>
  );
}
