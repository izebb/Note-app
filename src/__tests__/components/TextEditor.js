import React from "react";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { TextEditor } from "../../components/TextEditor";

describe("TextEditor", () => {
  it("should render without errors", () => {
    const props = {
      onEditorChange: jest.fn(),
      save: jest.fn(),
      textEditorInput: ""
    };
    const wrapper = mount(<TextEditor {...props} />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should call onEditorChange onchange and save on blur", () => {
    const props = {
      onEditorChange: jest.fn(),
      save: jest.fn(),
      textEditorInput: ""
    };

    const wrapper = mount(<TextEditor {...props} />);
    wrapper
      .find("textarea")
      .simulate("change", { target: { value: "new note" } })
      .simulate("blur");
    expect(props.onEditorChange).toHaveBeenCalled();
    expect(props.save).toHaveBeenCalled();
  });
});
