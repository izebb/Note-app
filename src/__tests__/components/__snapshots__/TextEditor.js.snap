// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextEditor should render without errors 1`] = `
<TextEditor
  onEditorChange={[MockFunction]}
  save={[MockFunction]}
  textEditorInput=""
>
  <TextEditor__EditorContainerStyled>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "TextEditor__EditorContainerStyled-pfhyks-0",
            "isStatic": true,
            "lastClassName": "cgLLHa",
            "rules": Array [
              "flex:1;",
            ],
          },
          "displayName": "TextEditor__EditorContainerStyled",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "TextEditor__EditorContainerStyled-pfhyks-0",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="TextEditor__EditorContainerStyled-pfhyks-0 cgLLHa"
      >
        <TextEditor__TextArea
          onBlur={[MockFunction]}
          onChange={[MockFunction]}
          value=""
        >
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [],
                "componentStyle": ComponentStyle {
                  "componentId": "TextEditor__TextArea-pfhyks-1",
                  "isStatic": true,
                  "lastClassName": "bghDmZ",
                  "rules": Array [
                    "width:100%;box-sizing:border-box;border:none;height:100%;background:transparent;font-size:16px;line-height:1.6;font-weight:300;padding:20px;white-space:pre-wrap;resize:none;&:focus{outline:none;}",
                  ],
                },
                "displayName": "TextEditor__TextArea",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "TextEditor__TextArea-pfhyks-1",
                "target": "textarea",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
            onBlur={[MockFunction]}
            onChange={[MockFunction]}
            value=""
          >
            <textarea
              className="TextEditor__TextArea-pfhyks-1 bghDmZ"
              onBlur={[MockFunction]}
              onChange={[MockFunction]}
              value=""
            />
          </StyledComponent>
        </TextEditor__TextArea>
      </div>
    </StyledComponent>
  </TextEditor__EditorContainerStyled>
</TextEditor>
`;
